racksdb/web/ui/src/components/InfrastructureTable.vue

```diff
@@ -10,41 +10,112 @@ import type {
+var popUpContent: Ref<
+  NodeEquipment | NetworkEquipment | StorageEquipment | MiscEquipment | undefined
+> = ref()
+const alphabeticalOrder = ref(true)
+const rackDisplayStates: Ref<Array<boolean>> = ref([])
+
+type EquipmentType = 'nodes' | 'storage' | 'network' | 'misc'
+interface Node extends NodeEquipment {
+  equipmentType: EquipmentType
+}
 
-// This function changes the display and filter the infrastructure to only get the data from a rack
-function rackDetails(rackName: string) {
-  showRack.value = true
-  showInfrastructureRacks.value = false
-  var layout = props.infrastructureDetails.layout
+interface Network extends NetworkEquipment {
+  equipmentType: EquipmentType
+}
+
+interface Storage extends StorageEquipment {
+  equipmentType: EquipmentType
+}
+
+interface Misc extends MiscEquipment {
+  equipmentType: EquipmentType
+}
 
-  for (let index = 0; index < layout!.length; index++) {
-    rack.value = layout!.filter((rack) => rack.rack === rackName)
+function displayRackEquipment(rackName: string) {
+  const index = rackNames.value.indexOf(rackName)
+  if (index !== -1) {
+    rackDisplayStates.value[index] = !rackDisplayStates.value[index]
   }
 }
 
-function allInfrastructurDetails() {
-  showRack.value = false
-  showInfrastructureRacks.value = true
+function sortRooms() {
+  if (rackNames.value) {
+    rackNames.value.reverse()
+  }
+  alphabeticalOrder.value = !alphabeticalOrder.value
+}
+
+function closePopUp() {
+  showPopUp.value = !showPopUp.value
+}
+
+function getRacks() {
+  var layout = props.infrastructureDetails.layout
+
+  layout.forEach((rack) => {
+    rackNames.value.push(rack.rack)
+  })
+}
+
+// This function get in parameter the name of the rack, filter the layout with the rack name and return the result
+function getRackEquipments(rackName: string) {
+  let equipments: Array<Node | Network | Storage | Misc> = []
+  const parts = props.infrastructureDetails.layout.filter((part) => part.rack == rackName)
+
+  /*
+   * For every parts, insert all nodes, network equipments, etc... enriched
+   * with equipmentType in equipments array.
+   */
+  parts.forEach((part) => {
+    equipments.push(
+      ...part.nodes.map((node) => {
+        return { ...{ equipmentType: 'nodes' }, ...node } as Node
+      })
+    )
+    equipments.push(
+      ...part.network.map((network) => {
+        return { ...{ equipmentType: 'network' }, ...network } as Network
+      })
+    )
+    equipments.push(
+      ...part.storage.map((storage) => {
+        return { ...{ equipmentType: 'storage' }, ...storage } as Storage
+      })
+    )
+    equipments.push(
+      ...part.misc.map((misc) => {
+        return { ...{ equipmentType: 'misc' }, ...misc } as Misc
+      })
+    )
+  })
+  // Sort equipment by slot in descending order
+  return equipments.sort((a, b) => b.slot - a.slot)
 }
 
-function popUp(name: string, equipment: equipmentType) {
+function popUp(name: string, equipmentType: 'nodes' | 'network' | 'storage' | 'misc') {
```

Comment:
Why don't you use `EquipmentType` type defined above?

Reply:
I did not think about using `EquipmentType`, but I made the change and used it.